brille/design/Proposed_integrated_syntax_Brill_SpinW_Horace.md

```diff
@@ -0,0 +1,267 @@
+```
+
+The above example assumes that data processing to construct an `.sqw` file combining multiple measurements of a crystal at different orientations is done.
+Triplets of numbers indicate that the data is to be rebinned along that axis (defined by the projection `proj`), whilst doublets indicates the data should be integrated between those ranges.
+Thus the example generates a 3D `sqw` object `cut1` from the 4D data in the `sqw_file`.
+
+The proposed syntax for Python is:
+
+<table>
+<td><img src="https://placehold.it/5x100/000000/000000?text=+"></td>
+<td>
+<pre lang="python">
+from pyHorace import horace as h
+proj = {'u':[1,0,0], 'v':[0,1,0], 'type':'rrr'}
+cut1 = h.cut_sqw(sqw_file, proj, [hmin, hstep, hmax], [kmin, kstep, kmax], 
+                                 [lmin, lmax], [emin, estep, emax])
+</pre>
+</td>
+<td><img src="https://placehold.it/5x100/000000/000000?text=+"></td>
+</table>
+
+For powder data, this is currently possible using the MSlice program.
+There are two implementations of MSlice: [one](https://github.com/mantidproject/mslice) in Python as part of [Mantid](https://www.mantidproject.org)
+and [the original](http://mslice.isis.rl.ac.uk/Main_Page) written in Matlab.
+The original Matlab MSlice is mostly a GUI program with an undocumented CLI (by using the underlying functions of the GUI).
+This program is also currently not being maintained so we will not use it.
+
+In Python, the syntax to extract data from a powder measurement using Mantid MSlice is:
+
+```python
+import mslice.cli as m
+dat2 = m.Load(nxs_file)
+cut2 = m.Slice(d2, f'|Q|,{qmin},{qmax},{qstep}', f'DeltaE,{emin},{emax},{estep}')
+```
+
+Note that unlike in Horace, the original pixel information is not retained in this rebinning operation.
+
+We propose the following syntax for powder data handling in Horace, should it be implemented:
+
+<table>
+<td><img src="https://placehold.it/5x70/000000/000000?text=+"></td>
+<td>
+<pre lang="matlab">
+proj = projaxes('powder');
+cut2 = h.cut_sqw(sqw_file, proj, [qmin, qstep, qmax], [emin, estep, emax]);
+</pre>
+</td>
+<td><img src="https://placehold.it/5x70/000000/000000?text=+"></td>
+</table>
+
+This syntax should retain pixel information just like the equivalent single crystal invocation.
+
+
+## Modelling the data
+
+We assume that [SpinW](https://spinw.org/) would be used to model the magnetic excitations.
+To set up and evaluate a model on single crystal data with current syntax in Matlab is:
+
+```matlab
+tri = sw_model('triAF', 1); 
+spinw_pars = {'hermit', false, 'formfact', true, 'resfun', 'gauss'}; 
+sim1 = disp2sqw_eval(cut1, @tri.horace_sqw, spinw_pars, fwhm);
+```
+
+This same syntax should also be possible (unchanged) for powder data, which has not been implemented.
+
+In Python it would be:
+
+<table>
+<td><img src="https://placehold.it/5x70/000000/000000?text=+"></td>
+<td>
+<pre lang="python">
+tri = h.sw_model('triAF', 1)
```

Comment:
Yeah, this is a bit tricky. 

I guess if we split things into different namespaces we should do something like:

```python
from pyPACE import pySpinW as sw
from pyPACE import pyHorace as h
```

where the big compiled library ("PACE") has separate namespaces for each program, but they are in the same Matlab compiled library and share the same interpreter / engine.

Which leads to the second "gotcha" which is that Matlab forbids the exchange of data between different Matlab interpreters / engines spawn by Python. This is even in the case where they might be in the same process and share memory (it's because the two engines are loaded from different `dll` and the libraries might have different ABI so that the memory layout is not compatible between them.) We can override this with some mex-file hackery but we would also have to ensure that if we distribute separate libraries for Horace and SpinW that the two libraries should be built by the same compiler with the same compiler options.

I agree that importing `meuphonic` (or whatever it's eventually called) into Python would not be a good idea, but the `pyPACE` library (or maybe "distribution"?) could also have the *Python* version of euphonic included instead.